cantools/database/can/cpp_source.py

```diff
@@ -0,0 +1,1037 @@
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/**
+ * This file was generated by cantools version {version} {date}.
+ */
+
+#ifndef {include_guard}
+#define {include_guard}
+
+#include "Frame.h"
+
+{declarations}
+
+#endif  // {include_guard}
+'''
+
+SOURCE_FMT = '''\
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018-2019 Erik Moqvist
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/**
+ * This file was generated by cantools version {version} {date}.
+ */
+
+#include "{header}"
+
+{definitions}\
+'''
+
+MESSAGE_DECLARATION_FMT = '''\
+/**
+ * Message {database_message_name}.
+ *
+{comment}\
+ */
+class {database_message_name} : public Frame {{
+public:
+    /** Constructor */
+    {database_message_name}(uint8_t* buffer, size_t size);
+
+    {database_message_name}& GetMessage() {{ 
+        return *this; 
+    }}
+ 
+    uint8_t* buffer() {{
+        return _buffer;
+    }}
+ 
+    size_t buffer_size() {{
+        return _buffer_size;
+    }}
+  
+    // Clear buffer
+    void clear() {{
+        memset(_buffer, 0, _buffer_size);
+    }}
+
+{members}
+
+private:
+    uint8_t* _buffer;
+    size_t _buffer_size;
+
+{encode_decode_members}
+}};
+'''
+
+SIGNAL_DECLARATION_FMT = '''\
+    /**
+     * Signal {name}.
+{comment}\
+     * Range: {range}
+     * Scale: {scale}
+     * Offset: {offset}
+{additional_comments}\
+     */
+    double {name}();
```

Comment:
This may not fit as well with the idioms provided by the underlying code generation, so take that suggestion with a grain of salt...

Reply:
This was in the first proposal I sent you over slack - debated this but thought it may be too much inheritance/back and forth. I do think it would be easier to work with as a consumer of the generated class.